libkbfs/tlf_handle_resolve.go

```diff
@@ -0,0 +1,595 @@
+		if conflictAdded {
+			// Can't add conflict info to a finalized handle.
+			return false, nil, TlfHandleFinalizedError{}
+		}
+	} else if other.IsFinal() {
+		finalizedAdded = true
+		// Ignore the added extension for resolution comparison purposes.
+		other.finalizedInfo = nil
+	}
+
+	unresolvedAssertions := make(map[string]bool)
+	for _, uw := range other.unresolvedWriters {
+		unresolvedAssertions[uw.String()] = true
+	}
+	for _, ur := range other.unresolvedReaders {
+		unresolvedAssertions[ur.String()] = true
+	}
+
+	// TODO: Once we keep track of the original assertions in
+	// TlfHandle, restrict the resolver to use other's assertions
+	// only, so that we don't hit the network at all.
+	partialResolvedH, err = h.ResolveAgain(
+		ctx, partialResolver{unresolvedAssertions, resolver})
+	if err != nil {
+		return false, nil, err
+	}
+
+	if conflictAdded || finalizedAdded {
+		resolvesTo, err = partialResolvedH.EqualsIgnoreName(codec, other)
+	} else {
+		resolvesTo, err = partialResolvedH.Equals(codec, other)
+	}
+	if err != nil {
+		return false, nil, err
+	}
+
+	return resolvesTo, partialResolvedH, nil
+}
+
+// MutuallyResolvesTo checks that the target handle, and the provided
+// `other` handle, resolve to each other.
+func (h TlfHandle) MutuallyResolvesTo(
+	ctx context.Context, codec kbfscodec.Codec,
+	resolver resolver, other TlfHandle, rev MetadataRevision, tlfID TlfID,
+	log logger.Logger) error {
+	handleResolvesToOther, partialResolvedHandle, err :=
+		h.ResolvesTo(ctx, codec, resolver, other)
+	if err != nil {
+		return err
+	}
+
+	// TODO: If h has conflict info, other should, too.
+	otherResolvesToHandle, partialResolvedOther, err :=
+		other.ResolvesTo(ctx, codec, resolver, h)
+	if err != nil {
+		return err
+	}
+
+	handlePath := h.GetCanonicalPath()
+	otherPath := other.GetCanonicalPath()
+	if !handleResolvesToOther && !otherResolvesToHandle {
+		return MDMismatchError{
+			rev, h.GetCanonicalPath(), tlfID,
+			fmt.Errorf(
+				"MD contained unexpected handle path %s (%s -> %s) (%s -> %s)",
+				otherPath,
+				h.GetCanonicalPath(),
+				partialResolvedHandle.GetCanonicalPath(),
+				other.GetCanonicalPath(),
+				partialResolvedOther.GetCanonicalPath()),
+		}
+	}
+
+	if handlePath != otherPath {
+		log.CDebugf(ctx, "handle for %s resolved to %s",
+			handlePath, otherPath)
+	}
+	return nil
+}
+
+type resolvableAssertionWithChangeReport struct {
+	resolvableAssertion
+	changed chan struct{}
+}
+
+func (ra resolvableAssertionWithChangeReport) resolve(ctx context.Context) (
+	nameUIDPair, keybase1.SocialAssertion, error) {
+	nuid, sa, err := ra.resolvableAssertion.resolve(ctx)
+	if err != nil {
+		return nuid, sa, err
+	}
+	if nuid.name.String() != "" {
+		if nuid.name.String() != ra.assertion {
+			sendIfPossible(ra.changed)
+		}
+	} else if sa != (keybase1.SocialAssertion{}) {
+		if sa.String() != ra.assertion {
+			sendIfPossible(ra.changed)
+		}
+	}
+	return nuid, sa, nil
+}
+
+func sendIfPossible(ch chan struct{}) {
```

Comment:
Any reason not to make this a function on `resolvableAssertionWithChangeReport` that doesn't need a channel passed in?  Seems like a weird place for a general function, and I can imagine wanting to use that name somewhere else in this library, possibly with a different signature.